pkg/providers/gcp/provider_postgres.go

```diff
@@ -3,63 +3,230 @@ package gcp
+		return croType.StatusMessage(errMsg), fmt.Errorf("%s: %w", errMsg, err)
 	}
-	// TODO implement me
-	return "", nil
+
+	// build cloudSQL service
+	sqladminService, err := sqladmin.NewService(ctx, option.WithCredentialsJSON(creds.ServiceAccountJson))
+	if err != nil {
+		return "error building cloudSQL admin service", err
+	}
+	sqlClient := &sqlClient{
+		sqlAdminService: sqladminService,
+	}
+	return pp.deleteCloudSQLInstance(ctx, sqlClient, p)
 }
 
-var _ providers.PostgresProvider = (*PostgresProvider)(nil)
+func (pp *PostgresProvider) deleteCloudSQLInstance(ctx context.Context, sqladminService SQLAdminService, p *v1alpha1.Postgres) (croType.StatusMessage, error) {
+	logger := pp.Logger.WithField("action", "deleteCloudSQLInstance")
+	// get cloudSQL instance
+	instances, err := getCloudSQLInstances(sqladminService)
+	if err != nil {
+		return "cannot retrieve sql instances from gcp", err
+	}
+	logrus.Info("listed sql instances from gcp")
+
+	instanceName := annotations.Get(p, ResourceIdentifierAnnotation)
+
+	if instanceName == "" {
+		msg := "unable to find instance name from annotation"
+		return croType.StatusMessage(msg), fmt.Errorf(msg)
+	}
+
+	// check if instance exists
+	var foundInstance *sqladmin.DatabaseInstance
+	for _, instance := range instances {
+		if instance.Name == instanceName {
+			logrus.Infof("found matching instance by name: %s", instanceName)
+			foundInstance = instance
+			break
+		}
+	}
+	// check instance state
+	if foundInstance != nil {
+		if foundInstance.State != "RUNNABLE" {
+			statusMessage := fmt.Sprintf("delete detected, DeletePostgres() is in progress, current cloudSQL status is %s", foundInstance.State)
+			pp.Logger.Info(statusMessage)
+			return croType.StatusMessage(statusMessage), nil
+		}
+		// delete if not in progress
+		_, err = sqladminService.DeleteInstance(ctx, projectID, instanceName)
+		if err != nil {
+			return croType.StatusMessage(fmt.Sprintf("failed to delete cloudSQL instance: %s", instanceName)), err
+		}
+		logrus.Info("triggered Instances.Delete()")
+		return "delete detected, Instances.Delete() started", nil
+	}
+
+	// delete credential secret
+	logger.Info("deleting cloudSQL secret")
+	sec := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      p.Name + defaultCredSecSuffix,
+			Namespace: p.Namespace,
+		},
+	}
+	err = pp.Client.Delete(ctx, sec)
+	if err != nil && !k8serr.IsNotFound(err) {
+		msg := "failed to deleted cloudSQL secrets"
+		return croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	// remove finalizer
+	resources.RemoveFinalizer(&p.ObjectMeta, DefaultFinalizer)
+	if err := pp.Client.Update(ctx, p); err != nil {
+		msg := "failed to update instance as part of finalizer reconcile"
+		return croType.StatusMessage(msg), errorUtil.Wrapf(err, msg)
+	}
+
+	return croType.StatusEmpty, nil
+}
+
+func getCloudSQLInstances(service SQLAdminService) ([]*sqladmin.DatabaseInstance, error) {
+	// check for existing instance
```

Comment:
This comment shouldn't be here as we're not checking for the existence of a particular instance.

Reply:
Removing comment.